examples/c/ex_storage_source.c

```diff
@@ -0,0 +1,1142 @@
+    }
+    unlock_storage_source(&demo_ss->lock);
+    if (ret == 0)
+        return (0);
+
+    (void)wtext->err_printf(wtext, session,
+      "%s: handle-write: failed to write %zu bytes at offset %zu: %s", demo_fh->iface.name, len,
+      off, wtext->strerror(wtext, NULL, ret));
+    return (ret);
+}
+
+/*
+ * demo_handle_remove --
+ *     Destroy an in-memory file handle. Should only happen on remove or shutdown.
+ */
+static int
+demo_handle_remove(WT_SESSION *session, DEMO_FILE_HANDLE *demo_fh)
+{
+    DEMO_STORAGE_SOURCE *demo_ss;
+    WT_EXTENSION_API *wtext;
+
+    demo_ss = demo_fh->demo_ss;
+    wtext = demo_ss->wtext;
+
+    if (demo_fh->ref != 0) {
+        (void)wtext->err_printf(wtext, session, "demo_handle_remove: %s: file is currently open",
+          demo_fh->iface.name, wtext->strerror(wtext, NULL, EBUSY));
+        return (EBUSY);
+    }
+
+    TAILQ_REMOVE(&demo_ss->fileq, demo_fh, q);
+
+    /* Clean up private information. */
+    free(demo_fh->buf);
+
+    /* Clean up public information. */
+    free(demo_fh->iface.name);
+
+    free(demo_fh);
+
+    return (0);
+}
+
+/*
+ * demo_handle_search --
+ *     Return a matching handle, if one exists.
+ */
+static DEMO_FILE_HANDLE *
+demo_handle_search(
+  WT_STORAGE_SOURCE *storage_source, WT_LOCATION_HANDLE *location_handle, const char *name)
+{
+    DEMO_FILE_HANDLE *demo_fh;
+    DEMO_STORAGE_SOURCE *demo_ss;
+    size_t len;
+    char *location;
+
+    demo_ss = (DEMO_STORAGE_SOURCE *)storage_source;
+    location = (char *)location_handle;
+    len = strlen(location);
+
+    TAILQ_FOREACH (demo_fh, &demo_ss->fileq, q)
+        if (strncmp(demo_fh->iface.name, location, len) == 0 &&
+          strcmp(&demo_fh->iface.name[len], name) == 0)
+            break;
+    return (demo_fh);
+}
+
+static const char *home;
+
+static int
+demo_test_create(WT_STORAGE_SOURCE *ss, WT_SESSION *session, WT_LOCATION_HANDLE *location,
+  const char *objname, const char *content)
+{
+    WT_FILE_HANDLE *fh;
+    const char *op;
+    size_t len;
+    int ret, t_ret;
+
+    fh = NULL;
+    len = strlen(content) + 1;
+    op = "open";
+    if ((ret = ss->ss_open_object(ss, session, location, objname, WT_SS_OPEN_CREATE, &fh)) != 0)
+        goto err;
+    op = "write";
+    if ((ret = fh->fh_write(fh, session, 0, len, content)) != 0)
+        goto err;
+
+err:
+    if (fh != NULL && (t_ret = fh->close(fh, session)) != 0 && ret == 0) {
+        op = "close";
+        ret = t_ret;
+    }
+    if (ret != 0)
+        fprintf(stderr, "demo failed during %s: %s\n", op, wiredtiger_strerror(ret));
+    else
+        printf("demo succeeded create %s\n", objname);
+
+    return (ret);
+}
+
+static int
+demo_test_read(WT_STORAGE_SOURCE *ss, WT_SESSION *session, WT_LOCATION_HANDLE *location,
+  const char *objname, const char *content)
+{
+    WT_FILE_HANDLE *fh;
+    char buf[100];
+    const char *op;
+    size_t len;
+    wt_off_t size;
+    int ret, t_ret;
+
+    fh = NULL;
+    len = strlen(content) + 1;
```

Comment:
Add a comment here before the first `op` saying  something about "set the op string so that on error we know what failed". I had a comment in here asking where it was used but then found it at the end.